### ISSUE TITLE
Moving caConstants & pixelTopology to traits

### ISSUE THREAD

**Issue by AdrianoDee** [cms-sw/cmssw#36809]
More than an issue this is to ask for advice. Being in the process of cleaning https://github.com/cms-sw/cmssw/pull/36235, adding the pixel track reco and being inspired by Andrea's comment, [here](https://github.com/cms-sw/cmssw/pull/36235#issuecomment-979778327) I was thinking about restructuring `caConstants` and `pixelTopology` to be traits/structs and then template data structures and classes (similarly to what is done for CPU/GPUTraits). Eventually also including the geometry constants under  `gpuPixelDoublets`. So one would have something like (e.g.) :

```C++ 
template <typename TTraits, typename TrackerTraits>
class CAHitNtupletGeneratorKernels {
public:
  using Traits = TTraits; 
  using Topology = TrackerTraits;
  
  ...
  
  using HitToTuple = HitToTupleT<TrackerTraits>;
  
  ...
  
  }

  ...

using CAHitNtupletGeneratorKernelsGPU = CAHitNtupletGeneratorKernels<cms::cudacompat::GPUTraits,somenamespace::phase1Tracker>;
using CAHitNtupletGeneratorKernelsCPU = CAHitNtupletGeneratorKernels<cms::cudacompat::CPUTraits,somenamespace::phase1Tracker>;
using CAHitNtupletGeneratorKernelsGPUPhase2 = CAHitNtupletGeneratorKernels<cms::cudacompat::GPUTraits,somenamespace::phase2Tracker>;
using CAHitNtupletGeneratorKernelsCPUPhase2 = CAHitNtupletGeneratorKernels<cms::cudacompat::CPUTraits,somenamespace::phase2Tracker>;
  ```
  
This would allow:
- to completely disentangle Phase-I and Phase-II via templates. Considered that data structures for Phase-II are quite larger;
- to eventually customize the CA (and the local reco) for different topologies in a (more) straightforward way.

One may have both the `pixelTopolgy` helpers/constants and `caConstants` structures/constants in the same struct. On this last point, as on the main one, any suggestion is welcome.

**Comment by AdrianoDee** [comment]
assign heterogeneous,reconstruction

**Comment by fwyzard** [comment]
As long as they can be used at compile time, I agree that encapsulating constants and simple methods in one or more "traits" `struct`s can be more flexible than `namespace`s: traits can be passed as template parameters (and can have a default), while `namespace`s cannot.

**Comment by AdrianoDee** [comment]
Thanks @fwyzard, I'll set up a PR to have something concrete to discuss about.

**Comment by AdrianoDee** [comment]
@jpata not yet. Will ping here as soon as I open it

**Comment by AdrianoDee** [comment]
This could be closed (https://github.com/cms-sw/cmssw/pull/38761 was merged).

**Comment by perrotta** [comment]
> This could be closed (#38761 was merged).
